docs/source/java/memory.rst

```diff
@@ -0,0 +1,189 @@
+
+* Use the BufferAllocator interface in APIs instead of RootAllocator.
+* Create one RootAllocator at the start of the program.
+* ``close()`` allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.
+
+Debugging Memory Leaks/Allocation
+=================================
+
+Allocators have a debug mode that makes it easier to figure out where a leak is originated.
+To enable it, enable assertions with ``-ea`` or set the system property, ``-Darrow.memory.debug.allocator=true``.
+When enabled, a log will be kept of allocations.
+
+Arrow logs some allocation information via SLF4J; configure it properly to see these logs (e.g. via Logback/Apache Log4j).
+
+Consider the following example to see how debug enabled help us with the tracking of allocators:
+
+.. code-block:: Java
+
+    import org.apache.arrow.memory.ArrowBuf;
+    import org.apache.arrow.memory.BufferAllocator;
+    import org.apache.arrow.memory.RootAllocator;
+
+    try(BufferAllocator bufferAllocator = new RootAllocator(8 * 1024)){
+        ArrowBuf arrowBuf = bufferAllocator.buffer(4 * 1024);
+        System.out.println(arrowBuf);
+    }
+
+Logs without debug allocators enabled:
+
+.. code-block:: shell
+
+    11:56:48.944 [main] INFO  o.apache.arrow.memory.BaseAllocator - Debug mode disabled.
+    ArrowBuf[2], address:140508391276544, length:4096
+    16:28:08.847 [main] ERROR o.apache.arrow.memory.BaseAllocator - Memory was leaked by query. Memory leaked: (4096)
+    Allocator(ROOT) 0/4096/4096/8192 (res/actual/peak/limit)
+
+Logs with debug allocators enabled:
+
+.. code-block:: shell
+
+    11:56:48.944 [main] INFO  o.apache.arrow.memory.BaseAllocator - Debug mode enabled.
+    ArrowBuf[2], address:140437894463488, length:4096
+    Exception in thread "main" java.lang.IllegalStateException: Allocator[ROOT] closed with outstanding buffers allocated (1).
+    Allocator(ROOT) 0/4096/4096/8192 (res/actual/peak/limit)
+      child allocators: 0
+      ledgers: 1
+        ledger[1] allocator: ROOT), isOwning: , size: , references: 1, life: 261438177096661..0, allocatorManager: [, life: ] holds 1 buffers.
+            ArrowBuf[2], address:140437894463488, length:4096
+      reservations: 0
+
+.. _`BufferAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/BufferAllocator.html
+.. _`RootAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/RootAllocator.html
+.. _`newChildAllocator`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/RootAllocator.html#newChildAllocator-java.lang.String-org.apache.arrow.memory.AllocationListener-long-long-
+.. _`ArrowBuf`: https://arrow.apache.org/docs/java/reference/org/apache/arrow/memory/ArrowBuf.html
+.. _`Reference Counting`: https://github.com/apache/arrow/blob/2092e18752a9c0494799493b12eb1830052217a2/java/memory/memory-core/src/main/java/org/apache/arrow/memory/ReferenceManager.java#L30
```

Comment:
Wait - I don't think we should link into source files. They don't really provide useful context in this tutorial. (That goes for all the links here.)

Reply:
Deleted